src/tree/identity_tree.rs

```diff
@@ -192,20 +256,27 @@ impl IdentityTree {
                 })
                 .collect::<Vec<_>>();
 
-            // Sort leaves by leaf idx
             leaf_updates.sort_by_key(|(idx, _)| *idx);
 
-            // Apply all leaf updates to the tree
-            for (leaf_idx, val) in leaf_updates {
-                // Insert/update leaves in the canonical tree
-                // Note that the leaves are inserted/removed from the leaves hashmap when the updates are first applied to tree_updates
-                if val == Hash::ZERO {
-                    self.tree.set_leaf(leaf_idx as usize, Hash::ZERO);
+            // Sort leaf updates into insertions and deletions
+            let num_updates = leaf_updates.len();
+            let mut insertions = Vec::with_capacity(num_updates);
+            let mut deletions = Vec::with_capacity(num_updates);
+            for (leaf_idx, value) in leaf_updates.into_iter() {
```

Comment:
Not sure if we care about perf here, but we we can make this a lot faster in parallel.